### ISSUE TITLE
URL Rewrite destroyed

### ISSUE THREAD

**Issue by bernd-reindl** [magento/magento2#26940]
### Preconditions (*)

1. Magento 2.3.4

### Steps to reproduce (*)

1. Setup a multi website (not store) environment. (E.G. Website 1 = DE; Website 2 = EN)
2. Add a categories and translate the url path (E.G. global:Kategorie 1 (kategorie-1); website 2:Category 1 (category-1)
3. Add a product to the category and assign only to website 1 (Do not assign the product to website 2)
4. Check the URL rewrites.
5. Go to the category and change the path of the category in website 2 (E.G. category-en)
6. Check the URL rewrites again

### Expected result (*)

1. The paths for product containing the categories should be german in website 1

### Actual result (*)

1. The paths for product containing the categories are english in website 1

### Description
When a category path was changed in website > 1 and a product is not assignt to this website, the paths for this product will be corrupted in the default website (1)

**Comment by bernd-reindl** [comment]
I can confirm this is reproduceable in vanilla Magento (2.4-dev)

![url_rewrite](https://user-images.githubusercontent.com/2369089/74921717-78023380-53ce-11ea-90ce-aa2bb9aadb0b.jpg)
![category-website 1](https://user-images.githubusercontent.com/2369089/74921738-7d5f7e00-53ce-11ea-8442-341006795e35.jpg)
![category-website 2](https://user-images.githubusercontent.com/2369089/74921764-83555f00-53ce-11ea-9206-2ad019f77ac4.jpg)

**Comment by bernd-reindl** [comment]
Issue: Clear Description
Reproduced on 2.4.x

**Comment by GDauer** [comment]
@magento give me 2.4-develop instance - upcoming 2.4.x release

**Comment by magento-engcom-team** [comment]
Hi @GDauer. Thank you for your request. I'm working on Magento 2.4-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @GDauer, here is your [Magento instance](https://i-26940-2-4-develop.instances.magento-community.engineering).
Admin access: [https://i-26940-2-4-develop.instances.magento-community.engineering/admin_1e8c](https://i-26940-2-4-develop.instances.magento-community.engineering/admin_1e8c)
Login: `0e7a6a93` Password: `7030602188fd`
Instance will be terminated in up to 3 hours.

**Comment by GDauer** [comment]
@magento give me 2.3.4 instance

**Comment by magento-engcom-team** [comment]
Hi @GDauer. Thank you for your request. I'm working on Magento 2.3.4 instance for you

**Comment by magento-engcom-team** [comment]
Hi @GDauer, here is your [Magento instance](https://i-26940-2-3-4.instances.magento-community.engineering).
Admin access: [https://i-26940-2-3-4.instances.magento-community.engineering/admin_34bb](https://i-26940-2-3-4.instances.magento-community.engineering/admin_34bb)
Login: `aaaf757c` Password: `637532920abf`
Instance will be terminated in up to 3 hours.

**Comment by GDauer** [comment]
Hi @bernd-reindl i'm trying to work on your issue but i could not reproduce it. Can you give more information, maybe a video?

**Comment by bernd-reindl** [comment]
i already have reproduced it in https://i-26940-2-4-develop.magento-testing-service.engineering/admin_afb8

See my screenshots above.

**Comment by bernd-reindl** [comment]
https://drive.google.com/file/d/17pbMZ6p_E9ok4HwfVlAv69z8eSS5FAOD/view?usp=sharing

Create 2 Websites. First DE, Second EN
Add a category but don´t translate for EN for now
Add a product and add it only to DE
Check the URL rewirtes for product in website DE
Change the path of category for website EN
Check the URL rewrites again

Expected result: URL should not changed, because product is only in website DE
Actual result: The URL has changed to category-1/produkt-1.html in website DE

If the product is assigned to website DE and EN, the URL for DE is kategorie-1/produkt-1.html and for EN category-1/product-1.html.
This only happens if the product is not assigned to all existing websites

Hope this helps.

**Comment by bernd-reindl** [comment]
Hi @GDauer, 

i have found a solution.

in Magento\Catalog\Model\ResourceModel\Product
in function getCategoryCollection()
the store ID from product is not set for category collection. So the returned categoies return url_key and url_path from selected store in the backend view. Not the store for whitch the url rewrite is generated.

So i have changed the code like this:
```
public function getCategoryCollection($product)
    {
        /** @var Category\Collection $collection */
        $collection = $this->_categoryCollectionFactory->create();
		if($product->getStoreId()) {
			$collection->setStoreId($product->getStoreId());
		}
        $collection->joinField(
            'product_id',
            'catalog_category_product',
            'product_id',
            'category_id = entity_id',
            null
        )->addFieldToFilter(
            'product_id',
            (int)$product->getEntityId()
        );
        return $collection;
    }

```

The second problem is in  Magento\CatalogUrlRewrite\Model\Product\AnchorUrlRewriteGenerator
in function generate()
the store ID isn't set when get the category in row 70. So i have changed

 ` $anchorCategory = $this->categoryRepository->get($anchorCategoryId);`
to 
 ` $anchorCategory = $this->categoryRepository->get($anchorCategoryId, $storeId);`

Now it works like expected.

**Comment by hostep** [comment]
@bernd-reindl: that last fix you mentioned is finally being added to the Magento core, it has been a bug since Magento2 got released, but they finally accepted the fix in https://github.com/magento/magento2/pull/26784
This fix will be included in Magento 2.4.0, but hopefully as well in Magento 2.3.5 or 2.3.6, but I'm not sure about that...
